README.md
FreshMail Subscription ![JED icon](./artwork/JED_icon.png "mod_freshmail2")
======================

[English documentation](https://github.com/piotr-cz/mod_freshmail2/blob/master/README.en.md)

Moduł zapisu do newslettera w systemie [FreshMail](http://freshmail.pl/)


Funkcje
-------
- Zarządzanie  polami dodatkowymi (Wyświetlane, Wymagane)
- Powiadomienie o nowym subksrybencie na email
- Opcja odnośnika do regulaminu w formularzu
- Układy (layouty) do wyboru
 - Horyzontalny (Bootstrap 2 / 3)
 - Rzędowy (Bootstrap 2 / 3)
 - Beez (Joomla 2.5)
 - Atomic (Joomla 2.5)
- Ograniczenie liczby wyświetleń lub ukrycie modułu po registracji
- Ajax (wymagana Joomla 3.1+)
- Wybór listy subkrypcyjnej
- Integracha Captcha


Wymagania
---------

- Joomla 2.5+ or 3.0+
- PHP 5.3
- Konto w systemie FreshMail


Instalacja
----------

1. Pobierz [najnowsze wydanie](https://github.com/piotr-cz/mod_freshmail2/releases/latest) i zainstaluj za pomocą Instalatora Rozszerzeń Joomla _(Rozszerzenia > Instalacje > Instaluj z pakietu)_.
2. Dodaj moduł _(Rozszerzenia > Moduły > Utwórz > Freshmail Subscription)_


Konfiguracja
------------

1. Wstaw **Klucz API (32 znaków)** konta Freshmail _(Opcje > Klucz API)_
2. Wstaw **API Sekret (40 znaków)** _(Opcje > API Sekret)_
3. Wybierz Listę odbiorców _(Opcje > Lista odbiorców)_


W Twoim [panelu freshmail.pl](https://app.freshmail.com/pl/dashboard/index/) Klucz API oraz API Sekret znajdziesz w _Odbiorcy > lista odbiorców > Parametry > Klucze API_ lub w _[Ustawienia > Wtyczki i API > Twoje klucze API](https://app.freshmail.com/pl/settings/integration/)_.


### Opcje podstawowe

![Basic options](./artwork/screenshots/screen-admin-opcje-podstawowe.png "Basic options")

### Opcje wzbogacone

![Advanced options](./artwork/screenshots/screen-admin-opcje-wzbogacone.png "Advanced options")

### Układy

**Układ wertykalny**

![Vertical layout](./artwork/screenshots/screen-site-wetykalny.png "Vertical layout")

**Układ Horyzontalny**

![Horizontal layout](./artwork/screenshots/screen-site-horyzontalny.png "Horizontal layout")


Układy wyboru listy
-------------------

Domyślny układ używa typu _checkox_. Aby zmienić na inny, zmień kod w układzie odpowiedzialny za wyświetlanie list:

**Radio**

```php
<?php // Lists:Control ?>
<?php foreach ($lists as $i => $list) : ?>
	<label class="radio" title="<?php echo $list->description ?>">
		<input name="<?php echo $control ?>[list][]" type="radio" <?php if ($list->selected) : ?> checked="checked"<?php endif ?> value="<?php echo $list->subscriberListHash ?>" />
		<?php echo $list->name ?>
	</label>
<?php endforeach ?>
```

**Select**

```php
<?php // Lists:Control ?>
<select name="<?php echo $control ?>[list]" style="width: 100%">
<?php foreach ($lists as $i => $list) : ?>
	<option value="<?php echo $list->subscriberListHash ?>" <?php if ($list->selected) : ?> selected="selected"<?php endif ?>><?php echo $list->name ?></option>
<?php endforeach ?>
</select>
```

**Checkbox**

```php
<?php // Lists:Control ?>
<?php foreach ($lists as $i => $list) : ?>
	<label class="radio" title="<?php echo $list->description ?>">
		<input name="<?php echo $control ?>[list]" type="radio" <?php if ($list->selected) : ?> checked="checked"<?php endif ?> value="<?php echo $list->subscriberListHash ?>" />
		<?php echo $list->name ?>
	</label>
<?php endforeach ?>
```


Własne szablony
---------------

Moduł komunikuje z użytkownikiem za pomocą wiadomości z wykorzystaniem 
[API Joomla](http://docs.joomla.org/Display_error_messages_and_notices).
W szablonie powinien więc znajdować się kod dla ich wyświetlania: 
`<jdoc:include type="message" />` jak np. w domyślnym szablonie Atomic.

Aby funkcjonalność AJAX działała we własnym szablonie, formularz powinien zawierać 
atrybut `data-freshmail2="<?php echo $control ?>"`.


Autorzy
-------

- [piotr_cz](https://github.com/piotr-cz)


Błędy/ Nowe funkcje
-------------------

[Zgłoś błąd lub propozycję nowej funkcji tutaj](https://github.com/piotr-cz/mod_freshmail2/issues)

[Najnowsza wersja developerska](https://github.com/piotr-cz/mod_freshmail2/archive/master.zip)


CHANGELOG.md
Change Log
==========
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

[Unreleased][unreleased]
------------------------

[v1.2.3] - 2018-06-11
### Added
- GDRP options

[v1.2.2] - 2018-05-25
---------------------
### Changed
- Do not use ajax functionality when messages container is not available
- Updated copyright to 2018

[v1.2.1] - 2016-10-09
---------------------
### Added
- Additional module security checks
- Added option to redirect to menu item on success

### Fixed
- Fixed closing elemnent tag in twbs23 layout

### Changed
- Rearranged module options

[v1.2.0] - 2016-07-18
---------------------
### Added
- Captcha support

### Fixed
- Fixed link to documentation (thanks @piotrslabik1)

[v1.1.7] - 2016-03-17
---------------------
### Fixed
- Conflict with email cloak plugin

[v1.1.6] - 2016-03-15
---------------------
### Changed
- More explicit message when API service is not available

[v1.1.5] - 2015-11-05
---------------------
### Changed
- Updated API Key and API Secret referencees to FreshMail 6.0
- Improvements to language strings

### Fixed
- Joomla 3.4 validation compatibility

[v1.1.4] - 2015-09-04
---------------------
### Fixed
- Updated FreshMail API URL as of 2015/08/04
- Codestyle fixes

[v1.1.3] - 2015-03-21
---------------------
### Added
- Remember form data when subscription fails
- Support for Bootstrap 3

### Fixed
- Ajax successful message
- PHP notice when using limits
- Using same cache folder for both Site and Admin
- String to array conversions

### Changed
- Subtle changes to Bootstrap layouts
- Updated FreshMail API URL as of 2015/02/22

[v1.1.2] - 2015-01-12
---------------------
### Changed
- Updating english language files

[v1.1.1] - 2014-12-15
----------------------
### Fixed
- Fixing installation bug (related to Removal of index.html files)

[v1.1.0] - 2014-10-27 [YANKED]
------------------------------
### Added
- Registration to multiple lists

### Changed
- Code refactor
- Removal of `index.html` files

[unreleased]: https://github.com/piotr-cz/mod_freshmail2/compare/v1.2.3...HEAD
[v1.2.3]: https://github.com/piotr-cz/mod_freshmail2/compare/v1.2.2...v1.2.3
[v1.2.2]: https://github.com/piotr-cz/mod_freshmail2/compare/v1.2.1...v1.2.2
[v1.2.1]: https://github.com/piotr-cz/mod_freshmail2/compare/v1.2.0...v1.2.1
[v1.2.0]: https://github.com/piotr-cz/mod_freshmail2/compare/v1.1.7...v1.2.0
[v1.1.7]: https://github.com/piotr-cz/mod_freshmail2/compare/v1.1.6...v1.1.7
[v1.1.6]: https://github.com/piotr-cz/mod_freshmail2/compare/v1.1.5...v1.1.6
[v1.1.5]: https://github.com/piotr-cz/mod_freshmail2/compare/v1.1.4...v1.1.5
[v1.1.4]: https://github.com/piotr-cz/mod_freshmail2/compare/v1.1.3...v1.1.4
[v1.1.3]: https://github.com/piotr-cz/mod_freshmail2/compare/v1.1.2...v1.1.3
[v1.1.2]: https://github.com/piotr-cz/mod_freshmail2/compare/v1.1.1...v1.1.2
[v1.1.1]: https://github.com/piotr-cz/mod_freshmail2/compare/v1.1.0...v1.1.1
[v1.1.0]: https://github.com/piotr-cz/mod_freshmail2/compare/v1.0.0-rc.3...v1.1.0


helper.php
<?php
/**
 * @package     Freshmail2.Site
 * @subpackage  mod_freshmail2
 *
 * @copyright   Copyright (C) 2013 - 2018 Piotr Konieczny. All rights reserved.
 * @license     GNU General Public License version 2 or later; see LICENSE.txt
 */

defined('_JEXEC') or die;

// Import dependencies
JLoader::register('JFmRestApi', __DIR__ . '/libraries/Freshmail/FmRestApi.php');

/**
 * Helper for mod_freshmail2
 *
 * @package     Freshmail.Site
 * @subpackage  mod_freshmail2
 * @since       1.0
 */
class ModFreshmail2Helper
{
	/**
	 * Check if can skip module rendering
	 *
	 * @param   string     $control  Form domain
	 * @param   JRegistry  $params   Extension parameters
	 *
	 * @return  boolean  True to skip module rendering
	 *
	 * @note    Doesn't use on cookie: Possible read/write conflict with
	 *          multiple modules as JInput doesn't throttle IO operations
	 *          till request EOL.
	 */
	public static function canSkip($control, JRegistry $params)
	{
		// Get parameters
		$limit_time = $params->get('limit_time', 0);
		$limit_count = $params->get('limit_count', 0);
		$limit_registered = $params->get('limit_registered', 0);

		// No cookies required
		if (!$limit_time && !$limit_count && !$limit_registered)
		{
			return false;
		}

		// Read cookie
		$inputCookie = JFactory::getApplication()->input->cookie;
		$dataString = $inputCookie->get('freshmail2_' . $control, null, 'string');

		// Decode data
		$data = ($dataString) ? json_decode($dataString) : static::getDefaultCookieObject();

		// Limit multiple registrations
		// TODO: Remove check for state after 2015-05-01 (change in cookies)
		if ($limit_registered && (!empty($data->subscribed) || !empty($data->state)))
		{
			return true;
		}

		// Check time limit against last shown
		if ($limit_time)
		{
			if ($data->time && $limit_time <= (time() - $data->time) / 60)
			{
				return true;
			}
			elseif (!$data->time)
			{
				$data->time = time();
			}
		}

		// Check count limit
		if ($limit_count)
		{
			if ($data->count && $limit_count <= $data->count)
			{
				return true;
			}
			else
			{
				++$data->count;
			}
		}

		// Store cookie data for with max-age of 1 month
		$inputCookie->set('freshmail2_' . $control, json_encode($data), time() + 30 * 24 * 3600);

		return false;
	}

	/**
	 * Get menu link by Id
	 *
	 * @param   integer  $itemId  Menu item ID
	 *
	 * @return  string  Menu item URI
	 */
	public static function getMenuLink($itemId = null)
	{
		$menu 	= JFactory::getApplication()->getMenu();
		$item	= $menu->getItem((int) $itemId);

		if (!$item)
		{
			return null;
		}

		return $item->link . '&Itemid=' . $itemId;
	}

	/**
	 * Execute Rest API command wrapper
	 *
	 * @param   string   $apiKey          FreshMail API Key
	 * @param   string   $apiSecret       FreshMail API Secret
	 * @param   string   $command         Command name
	 * @param   array    $params          Command parameters
	 * @param   string   $key             Result set key
	 * @param   boolean  $cacheCheckTime  Check cache time (Site operates on data updated by Admin)
	 *
	 * @return  array  Result set
	 *
	 * @throws  RestException
	 * @throws  Exception
	 */
	public static function executeCommand($apiKey, $apiSecret, $command, $params = array(), $key = null, $cacheCheckTime = true)
	{
		// Initialize cache
		$cache = JCache::getInstance(
			/* @type string $type */
			'output',
			/* @type array $options */
			array(
				'language'		=> 'en-GB',
				'defaultgroup' 	=> 'mod_freshmail2',
				'checkTime'		=> $cacheCheckTime,
				'caching'		=> true,
				// Use same admin cache folder for both admin and site
				'cachebase'		=> JPATH_ADMINISTRATOR . '/cache',
			)
		);

		// Build cache key
		$cacheId = $apiKey . ':' . $command . ':' . serialize($params);

		// Load response from cache
		$response = $cache->get($cacheId);

		// No cached items
		if ($response === false)
		{
			$client = new JFmRestApi;
			$client->setApiKey($apiKey);
			$client->setApiSecret($apiSecret);

			if (method_exists($client, 'setTimeout'))
			{
				$client->setTimeout(30);
			}

			/* @throws RestException */
			$client->doRequest($command, (is_array($params) ? $params : null));

			// Get Items from response
			$response = $client->getResponse();

			// Store response in cache
			$cache->store($response, $cacheId);
		}

		// Access result set by key
		if ($key && isset($response[$key]))
		{
			return $response[$key];
		}

		return $response;
	}

	/**
	 * Get and prepare custom fields
	 * [frontend]
	 *
	 * @param   JRegistry  $params  Module parameters
	 * @param   array      $values  Values [Optional]
	 *
	 * @return  array
	 */
	public static function getProcessedCustomFields(JRegistry $params, $values = array())
	{
		$items = array();
		$lists = (array) $params->get('FMlistHash');

		$tags = array();
		$uniqueFields = array();
		$allFields = array();

		// Item: hash, name, tag, type
		foreach ($lists as $listHash)
		{
			try
			{
				$allFields = array_merge(
					$allFields,
					static::executeCommand(
						$params->get('FMapiKey'),
						$params->get('FMapiSecret'),
						'subscribers_list/getFields',
						array('hash' => $listHash),
						'fields',
						false
					)
				);
			}
			catch (Exception $e)
			{
				return $items;
			}
		}

		// Save from pivot fatal
		if (empty($allFields))
		{
			return $items;
		}

		// Pivot by unique tag
		// Note: these are for display puproses only, there may be fields with same tag, accross multiple lists so latter will overwrite first.
		foreach ($allFields as $field)
		{
			if (!in_array($field['tag'], $tags))
			{
				$uniqueFields[$field['tag']] = $field;
			}
		}

		// Get params
		$displayFields = $params->get('FMdisplayFields', array());
		$requiredFields = $params->get('FMrequiredFields', array());

		// Filter in only display fields
		$uniqueFields = (array_intersect_key($uniqueFields, array_flip($displayFields)));

		// Add each item to collection
		foreach ($uniqueFields as $tag => $field)
		{
			// Add extra item data
			$field['required'] = (in_array($tag, $requiredFields));
			$field['value'] = (isset($values[$tag])) ? $values[$tag] : null;

			$items[] = (object) $field;
		}

		return $items;
	}

	/**
	 * Get and prepare subscribers list
	 * [frontend]
	 *
	 * @param   JRegistry  $params  Module parameters
	 * @param   array      $values  Values [Optional]
	 *
	 * @return  array
	 *
	 * @since   1.1
	 */
	public static function getProcessedLists(JRegistry $params, array $values = array())
	{
		$items = array();

		// Retrieve lists
		try
		{
			$allLists = static::executeCommand(
				$params->get('FMapiKey'),
				$params->get('FMapiSecret'),
				'subscribers_list/lists',
				null,
				'lists'
			);
		}
		catch (Exception $e)
		{
			return $items;
		}

		// Save from pivot fatal
		if (empty($allLists))
		{
			return $list;
		}

		// Pivot by listHash
		$allLists = JArrayHelper::pivot($allLists, 'subscriberListHash');

		// Get params
		$displayLists = (array) $params->get('FMlistHash');

		// Filter in only display data
		$allLists = (array_intersect_key($allLists, array_flip($displayLists)));

		$isSingleList = (count($allLists) == 1);

		// Add each item to collection
		foreach ($allLists as $hash => $list)
		{
			// Add extra item data
			$list['selected'] = ($isSingleList)
				? true
				: in_array($hash, $values);

			$items[] = (object) $list;
		}

		return $items;
	}

	/**
	 * Ajax event
	 * Ajaxified version of mod_freshmail2
	 *
	 * @return   mixed
	 *
	 * @note   Example use: ?option=com_ajax
	 *                      &format=json
	 *                      &module=freshmail2
	 *                      &method=post
	 *                      &ignoreMessages=0
	 *                      &control=[control]
	 *                      &[control]=[form]
	 *
	 * @note   Need javascript client to work.
	 *
	 * @since  3.1
	 */
	public static function postAjax()
	{
		// Get JInput object
		$app = JFactory::getApplication();
		$input = $app->input;

		// Load module language file
		$lang = JFactory::getLanguage();
		$lang->load('mod_freshmail2', JPATH_SITE)
			|| $lang->load('mod_freshmail2', __DIR__);

		// Get module id
		$control = $input->get('control');

		if (!$control)
		{
			return new LogicException('No module id');
		}

		// Load module instance
		$moduleId = substr($control, 4);
		$table = JTable::getInstance('module');

		if (!$table->load(array('id' => $moduleId)))
		{
			new LogicException(JText::sprintf('COM_AJAX_MODULE_NOT_ACCESSIBLE', $control), 404);
		}

		// Check published state and access levels
		if ($table->published != 1 || !in_array($table->access, JFactory::getUser()->getAuthorisedViewLevels()))
		{
			return new RuntimeException(JText('JERROR_ALERTNOAUTHOR'), 403);
		}

		// Check publish_up and publish_down
		$now = JFactory::getDate()->toSql();
		$dbNullDate = JFactory::getDbo()->getNullDate();
		$dbDateFormat = JFactory::getDbo()->getDateFormat();

		if (($table->publish_up != $dbNullDate && $now < DateTime::createFromFormat($dbDateFormat, $table->publish_up, 'UTC'))
			|| ($table->publish_down != $dbNullDate && $now > DateTime::createFromFormat($dbDateFormat, $table->publish_up, 'UTC'))
		)
		{
			return new RuntimeException(JText('JERROR_ALERTNOAUTHOR'), 403);
		}

		// Decode module params
		$params = new JRegistry;
		$params->loadString($table->params);

		// Read POSTed data
		$inputData = $input->post->get($control, null, 'array');

		// Get processed lists set
		$selectedLists = (isset($inputData['list'])) ? (array) $inputData['list'] : array();
		$lists = static::getProcessedLists($params, $selectedLists);

		// Process POSTed data (Valid, Added, Notified)
		if (!empty($inputData) && static::validate($inputData, $params))
		{
			// Success flag
			$addContactsSuccess = false;

			// Loop trough lists and process selected ones
			foreach ($lists as $list)
			{
				if ($list->selected)
				{
					// Validate, add contact
					if (!static::addContact($inputData, $params, $list))
					{
						$addContactsSuccess = false;
						break;
					}

					$addContactsSuccess = true;
				}
			}

			// Run post hooks
			if ($addContactsSuccess)
			{
				static::sendEmail($inputData, $params)
					&& static::postHook($control, $params);

				$response = array(
					'message' => JText::_('MOD_FRESHMAIL2_SUCCESS')
				);

				if ($params->get('redirectMenuitem'))
				{
					$menu = $app->getMenu();
					$menuItem = $menu->getItem($params->get('redirectMenuitem'));

					$response['redirectUrl'] = JRoute::_($menuItem->link . '&Itemid=' . $menuItem->id);
				}

				// All is OK
				return $response;
			}
		}

		return new UnexpectedValueException('Cannot subscribe');
	}

	/**
	 * Validate user input
	 *
	 * @param   array      $data    Data to validate
	 * @param   JRegistry  $params  Module parameters
	 *
	 * @return  boolean  True on success false on failure
	 */
	public static function validate(array $data, JRegistry $params)
	{
		$app = JFactory::getApplication();

		// CRSF token
		if (!JSession::checkToken('post'))
		{
			$app->enqueueMessage(JText::_('JINVALID_TOKEN'), 'error');

			return false;
		}

		// Captcha
		$captcha = static::getCaptcha($params);

		if ($captcha instanceof JCaptcha
			&& !$captcha->checkAnswer(isset($data['captcha']) ? $data['captcha'] : null)
		)
		{
			$error = $captcha->getError();

			$app->enqueueMessage($error instanceof Exception ? $error->getMessage() : $error, 'error');

			return false;
		}

		// Terms of Service
		if ($params->get('tos_menuitem') && empty($data['tos']))
		{
			$app->enqueueMessage(JText::_('MOD_FRESHMAIL2_ERROR_TOS'), 'error');

			return false;
		}

		// Email
		if (!JMailHelper::isEmailAddress($data['email']))
		{
			$app->enqueueMessage(JText::_('MOD_FRESHMAIL2_ERROR_EMAIL'), 'error');

			return false;
		}

		// In case of multiple lists, check at least one is selected
		// Note: should check set lists against submitted ones
		if (count((array) $params->get('FMlistHash')) > 1)
		{
			if (empty($data['list']))
			{
				$app->enqueueMessage(JText::_('MOD_FRESHMAIL2_ERROR_LIST_NOT_SELECTED'), 'error');

				return false;
			}
		}

		// Validate custom fields
		$validates = true;

		// Load up custom fields
		$customFields = static::getProcessedCustomFields($params);

		foreach ($customFields as $field)
		{
			if ($field->required && empty($data['custom_fields'][$field->tag]))
			{
				$app->enqueueMessage(JText::sprintf('MOD_FRESHMAIL2_ERROR_FIELD', $field->name), 'notice');

				// Invalidate
				$validates = false;
			}
		}

		return $validates;
	}

	/**
	 * Add contact to list
	 *
	 * @param   array      $data    Data to add
	 * @param   JRegistry  $params  Module parameters
	 * @param   stdClass   $list    List object
	 *
	 * @return  boolean  True on success false on failure
	 */
	public static function addContact(array $data, JRegistry $params, stdClass $list)
	{
		$app = JFactory::getApplication();
		$lang = JFactory::getLanguage();

		// Process messages differently
		$isSingle = (count($params->get('FMlistHash')) == 1);

		// Instanitate Client
		$client = new JFmRestApi;

		$client->setApiKey($params->get('FMapiKey'));
		$client->setApiSecret($params->get('FMapiSecret'));

		// Set timeout
		if (method_exists($client, 'setTimeout'))
		{
			$client->setTimeout($params->get('FMapiTimeout'));
		}

		/* Build payload
		 * Note: When 'confirm' value is set to null, Double opt-in lists
		 * don't send auto confirmations.
		 * Inconsistency in API docs: http://freshmail.pl/developer-api/subskrybenci-zarzadzanie-subskrybentami/
		 */
		$payload = array(
			'email'			=> $data['email'],
			'list'			=> $list->subscriberListHash,
			'custom_fields'	=> array(),
		//	'state'			=> null,
		//	'confirm'		=> null,
		);

		// Set Kody Statusow Subskrybentow
		if ($params->get('FMdefaultState'))
		{
			$payload['state'] = $params->get('FMdefaultState');
		}

		// Set Confirmation email
		if ($params->get('FMdefaultConfirm'))
		{
			$payload['confirm'] = $params->get('FMdefaultConfirm');
		}

		// Add custom fields
		// In payload, just custom field tag is required (not hash)
		if (isset($data['custom_fields']))
		{
			foreach ($params->get('FMdisplayFields', array()) as $customField)
			{
				// If filled in, set using tag as key
				if (isset($data['custom_fields'][$customField]))
				{
					$payload['custom_fields'][$customField] = $data['custom_fields'][$customField];
				}
			}
		}

		// Send payload
		try
		{
			/* @throws RestException */
			$client->doRequest('subscriber/add', $payload);
		}
		catch (Exception $e)
		{
			// Pop translated exception message if available
			$langKey = 'MOD_FRESHMAIL2_ERROR_' . $e->getCode();
			$message = ($lang->hasKey($langKey)) ? JText::_($langKey) : $e->getMessage();

			// Attach list name
			if (!$isSingle)
			{
				$message .= sprintf(' (%s)', $list->name);
			}

			$app->enqueueMessage($message, 'error');

			return false;
		}

		$response = $client->getResponse();

		// OK
		if ($response['status'] === 'OK')
		{
			if ($isSingle)
			{
				$app->enqueueMessage(JText::_('MOD_FRESHMAIL2_SUCCESS'), 'success');
			}

			return true;
		}

		// Undefined
		if ($response['status'] != 'ERROR')
		{
			return null;
		}

		// Render error message
		foreach ($response['errors'] as $error)
		{
			// Pop translated error message if available
			$langKey = 'MOD_FRESHMAIL2_ERROR_' . $error['code'];
			$message = ($lang->hasKey($langKey)) ? JText::_($langKey) : $error['message'];

			// Attach list name
			if (!$isSingle)
			{
				$message .= sprintf(' (%s)', $list->name);
			}

			$app->enqueueMessage($message, 'error');
		}

		return false;
	}

	/**
	 * Send notification email in enabled.
	 *
	 * @param   array      $data    Data to send
	 * @param   JRegisytr  $params  Module parameters
	 *
	 * @return  boolean  True on success false on failure
	 */
	public static function sendEmail(array $data, JRegistry $params)
	{
		// Initialise variables
		$app = JFactory::getApplication();
		$mailer = JFactory::getMailer();

		// Nofication sent to
		$to = $params->get('notificationTo');

		// Notification disabled
		if (!$params->get('notificationOn') || empty($to))
		{
			return true;
		}

		$mailFrom = $app->getCfg('mailfrom');
		$fromName = $app->getCfg('fromname');
		$siteName = $app->getCfg('sitename');

		// Clean email data
		$from		= JMailHelper::cleanAddress($data['email']);
		$subject	= JMailHelper::cleanSubject(JText::_('MOD_FRESHMAIL2_NOTIFICATION_SUBJECT'));

		// Prepare email body
		$body 		= "\r\n" . JText::sprintf('MOD_FRESHMAIL2_NOTIFICATION_BODY', $from);

		// Attach custom fields
		if (isset($data['custom_fields']))
		{
			$customFields = static::getProcessedCustomFields($params, $data['custom_fields']);

			foreach ($customFields as $field)
			{
				$body	.= "\r\n" . JText::sprintf('MOD_FRESHMAIL2_NOTIFICATION_FIELD', $field->name, $field->value);
			}
		}
		elseif (isset($data['lists']))
		{
			$lists = static::getProcessedLists($params, (array) $data['lists']);
			$body .= "\r\n" . JText::_('MOD_FRESHMAIL2_NOTIFICATION_LISTS');

			foreach ($lists as $list)
			{
				$body	.= "\r\n" . $list->name;
			}
		}

		// Construct mailer
		$mailer
			->addRecipient($to)
			->addReplyTo(array($from))
			->setSender(array($mailFrom, $fromName))
			->setSubject($siteName . ': ' . $subject)
			->setBody($body);

		return $mailer->Send();
	}

	/**
	 * Post submit hook
	 *
	 * @param   string     $control  Form domain
	 * @param   JRegistry  $params   Extension parameters
	 *
	 * @return  Boolean  True
	 */
	public static function postHook($control, JRegistry $params)
	{
		// Save state in cookie
		if ($params->get('limit_registered', 0))
		{
			$inputCookie = JFactory::getApplication()->input->cookie;
			$dataString = $inputCookie->get('freshmail2_' . $control, null, 'string');

			$data = ($dataString) ? json_decode($dataString) : static::getDefaultCookieObject();
			$data->subscribed = true;

			$inputCookie->set('freshmail2_' . $control, json_encode($data), time() + 30 * 24 * 3600);
		}

		return true;
	}

	/**
	 * Redirect
	 *
	 * @param   JRegistry  $params
	 *
	 * @return  void
	 */
	public static function redirect(JRegistry $params)
	{
		$app = JFactory::getApplication();
		$menu = $app->getMenu();
		$menuItem = $menu->getItem($params->get('redirectMenuitem'));

		$app->redirect(JRoute::_($menuItem->link . '&Itemid=' . $menuItem->id));
	}

	/**
	 * Get default cookie object
	 *
	 * @return stdClass
	 */
	public static function getDefaultCookieObject()
	{
		return (object) array(
			/* @type integer Timestamp of last access */
			'time' => 0,
			/* @type integer Module implressions count */
			'count' => 0,
			/* type boolean Used this module to subscribe */
			'subscribed' => false,
		);
	}

	/**
	 * Get captcha instance or null if not available
	 *
	 * @param   JRegistry  $params  Extension parameters
	 *
	 * @return  JCaptcha|null
	 */
	public static function getCaptcha(JRegistry $params)
	{
		$pluginName = $params->get('captcha', JFactory::getApplication()->get('captcha', '0'));

		// Disabled
		if ($pluginName == '0' || !$pluginName)
		{
			return null;
		}

		$plugins = JPluginHelper::getPlugin('captcha');

		// Selected plugin not available
		if (!in_array($pluginName, JArrayHelper::getColumn($plugins, 'name')))
		{
			return null;
		}

		return JCaptcha::getInstance($pluginName, array('namespace' => 'mod_freshmail2'));
	}
}
